### PR TITLE
feat(driver-adapters): First third party driver adapter from TiDB Cloud

### PR DESCRIPTION
**Pull request by janpio**

### CODE REVIEW

**Comment by jkomyno** [comment]
TODO:
- [ ] test once https://github.com/prisma/prisma/pull/21642 is merged
- [ ] port current one-shot executable code to actual tests?

**Comment by jkomyno** [comment]
Note: source code for `@tidbcloud/prisma-adapter` is available [here](https://www.npmjs.com/package/@tidbcloud/prisma-adapter?activeTab=code) (look for `./dist/index.mjs`).

File: driver-adapters/tidb-node-basic/index.test.js [hunk 1]
```diff
@@ -0,0 +1,107 @@
+// @ts-check
+const { test, expect } = require('@jest/globals')
+const { handler } = require('./index')
+const { dependencies } = require('./package.json')
+
+jest.setTimeout(10_000)
```
Comment (jkomyno) [inline driver-adapters/tidb-node-basic/index.test.js:6]: Is a custom timeout necessary?

**Review by jkomyno — APPROVED** [review]

File: driver-adapters/tidb-node-basic/index.test.js (hunk 1 above)
Comment (janpio) [reply driver-adapters/tidb-node-basic/index.test.js:6]: Copy/paste from the other driver-adapter tests 🤷

Comment (millsp) [reply driver-adapters/tidb-node-basic/index.test.js:6]: It is necessary as we are issuing many requests one after another and that may take more than the default 5 seconds.

File: driver-adapters/tidb-node-basic/package.json [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  },
+  "devDependencies": {
+    "@jest/globals": "29.7.0",
+    "jest": "28.1.3",
+    "prisma": "5.5.0-dev.58"
+  },
+  "dependencies": {
+    "@prisma/client": "5.5.0-dev.58",
+    "@tidbcloud/prisma-adapter": "^5.4.2",
+    "@tidbcloud/serverless": "^0.0.6"
```
Comment (millsp) [inline driver-adapters/tidb-node-basic/package.json]:
```suggestion
    "@tidbcloud/prisma-adapter": "5.4.2",
    "@tidbcloud/serverless": "0.0.6"
```

File: driver-adapters/tidb-node-basic/prisma/schema.prisma [hunk 3]
```diff
@@ -0,0 +1,17 @@
+generator client {
+  provider = "prisma-client-js"
+  previewFeatures = ["driverAdapters"]
+}
+
+datasource db {
+  provider = "mysql"
+  // TODO: needs to be changed to a fake url to ensure the driver adapter is used
+  url      = env("DRIVER_ADAPTERS_TIDB_NODE_BASIC_DATABASE_URL")
```
Comment (millsp) [inline driver-adapters/tidb-node-basic/prisma/schema.prisma]: Should this be a directUrl and a fake url instead?

File: driver-adapters/tidb-node-basic/run.sh [hunk 4]
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
+
+set -eu
+
+pnpm install
+pnpm prisma generate
```
Comment (millsp) [inline driver-adapters/tidb-node-basic/run.sh:6]: Should we invoke `prisma db push --force-reset` here since it won't be handled by CI automatically?

**Review by millsp — APPROVED** [review]

File: driver-adapters/tidb-node-basic/prisma/schema.prisma (hunk 3 above)
Comment (Jolg42) [inline driver-adapters/tidb-node-basic/prisma/schema.prisma]:
```suggestion
  // fake url to ensure the driver adapter is used
  url      = "mysql://foo"
  directUrl = env("DRIVER_ADAPTERS_TIDB_NODE_BASIC_DATABASE_URL")
```

File: driver-adapters/tidb-node-basic/run.sh (hunk 4 above)
Comment (Jolg42) [reply driver-adapters/tidb-node-basic/run.sh:6]: Since the workflow is already passing the test, it means the db is already in sync

**Review by Jolg42 — APPROVED** [review]
Nice!
See suggestion for fake url in schema

File: driver-adapters/tidb-node-basic/run.sh (hunk 4 above)
Comment (janpio) [reply driver-adapters/tidb-node-basic/run.sh:6]: I think this is just copy-pasted from the other tests - and seems to rely on the database to have been set up manually indeed.

I am ok with that.

File: driver-adapters/tidb-node-basic/package.json (hunk 2 above)
Comment (janpio) [reply driver-adapters/tidb-node-basic/package.json]: This broke the lockfile.

File: .github/workflows/test.yaml [hunk 5]
```diff
@@ -1251,6 +1252,8 @@ jobs:
       DRIVER_ADAPTERS_TURSO_VERCEL_NEXTJS_DATABASE_URL: ${{ vars.DRIVER_ADAPTERS_TURSO_VERCEL_NEXTJS_DATABASE_URL }}
       DRIVER_ADAPTERS_TURSO_VERCEL_NEXTJS_TOKEN: ${{ secrets.DRIVER_ADAPTERS_TURSO_VERCEL_NEXTJS_TOKEN }}
       DRIVER_ADAPTERS_TURSO_VERCEL_NEXTJS_PROJECT_ID: ${{ vars.DRIVER_ADAPTERS_TURSO_VERCEL_NEXTJS_PROJECT_ID }}
+      # TiDB Cloud
+      DRIVER_ADAPTERS_TIDB_NODE_BASIC_DATABASE_URL: ${{ secrets.DRIVER_ADAPTERS_TIDB_NODE_BASIC_DATABASE_URL }}
```
Comment (Jolg42) [inline .github/workflows/test.yaml:1256]: Is this set correctly? It looks like it fails the test because of this value

**Comment by janpio** [comment]
Currently blocked by https://github.com/prisma/team-orm/issues/662#issuecomment-1848371449

File: driver-adapters/tidb-node-basic/prisma/schema.prisma [hunk 6]
```diff
@@ -0,0 +1,18 @@
+generator client {
+  provider = "prisma-client-js"
+  previewFeatures = ["driverAdapters"]
+}
+
+datasource db {
+  provider = "mysql"
+  // fake url to ensure the driver adapter is used // TODO remove hardcoded database name here that currently is needed because of https://github.com/prisma/team-orm/issues/662
+  url      = "mysql://foo/test"
```
Comment (janpio) [inline driver-adapters/tidb-node-basic/prisma/schema.prisma]:
```suggestion
  // fake url to ensure the driver adapter is used
  url      = env("INVALID_ENV_VAR")
  // not used by the client, just to setup the db
```

File: driver-adapters/tidb-node-basic/prisma/schema.prisma [hunk 7]
```diff
@@ -0,0 +1,20 @@
+generator client {
+  provider = "prisma-client-js"
+  previewFeatures = ["driverAdapters"]
+}
+
+datasource db {
+  provider = "mysql"
+  // fake url to ensure the driver adapter is used // TODO remove hardcoded database name here that currently is needed because of https://github.com/prisma/team-orm/issues/662
```
Comment (janpio) [inline driver-adapters/tidb-node-basic/prisma/schema.prisma]:
```suggestion
```

**Pull request merged**